lib/concurrent/struct/final_struct.rb

```diff
@@ -0,0 +1,122 @@
+require 'concurrent/struct/abstract_struct'
+require 'concurrent/errors'
+require 'concurrent/synchronization'
+
+module Concurrent
+
+  # An thread-safe, write-one variation of Ruby's standard `Struct`.
+  # Each member can have its value set at most one, either at construction
+  # or any time thereafter. Attempting to assign a value to a member
+  # that has already been set will result in a `Concurrent::ImmutabilityError`.
+  #
+  # @see http://ruby-doc.org/core-2.2.0/Struct.html Ruby standard library `Struct`
+  # @see http://en.wikipedia.org/wiki/Final_(Java) Java `final` keyword
+  module FinalStruct
+    include AbstractStruct
+
+    # @!macro struct_values
+    def values
+      synchronize { ns_values }
+    end
+    alias_method :to_a, :values
+
+    # @!macro struct_values_at
+    def values_at(*indexes)
+      synchronize { ns_values_at(indexes) }
+    end
+
+    # @!macro struct_inspect
+    def inspect
+      synchronize { ns_inspect }
+    end
+    alias_method :to_s, :inspect
+
+    # @!macro struct_to_h
+    def to_h
+      synchronize { ns_to_h }
+    end
+
+    # @!macro struct_get
+    def [](member)
+      synchronize { ns_get(member) }
+    end
+
+    # @!macro struct_equality
+    def ==(other)
+      synchronize { ns_equality(other) }
+    end
+
+    # @!macro struct_each
+    def each
+      return enum_for(:each) unless block_given?
+      synchronize { ns_each(&Proc.new) }
```

Comment:
I did few benchmarks. MRI 2.2:
```txt
Calculating -------------------------------------
  def e1; yield; end    69.720k i/100ms
def e2(&block); block.call; end
                        43.656k i/100ms
-------------------------------------------------
  def e1; yield; end      3.782M (±20.5%) i/s -     34.581M
def e2(&block); block.call; end
                        869.372k (±17.0%) i/s -      8.295M

Comparison:
  def e1; yield; end:  3782497.8 i/s
def e2(&block); block.call; end:   869371.9 i/s - 4.35x slower

Calculating -------------------------------------
def e3; e1 &Proc.new; end
                        40.549k i/100ms
def e4(&block); e1 &block; end
                        40.700k i/100ms
-------------------------------------------------
def e3; e1 &Proc.new; end
                        806.050k (±16.8%) i/s -      7.704M
def e4(&block); e1 &block; end
                        884.341k (±16.0%) i/s -      8.547M

Comparison:
def e4(&block); e1 &block; end:   884341.4 i/s
def e3; e1 &Proc.new; end:   806049.8 i/s - 1.10x slower
```
JRuby 1.7.19
```txt
Calculating -------------------------------------
  def e1; yield; end   120.512k i/100ms
def e2(&block); block.call; end
                       116.150k i/100ms
-------------------------------------------------
  def e1; yield; end      4.408M (±15.4%) i/s -     42.420M
def e2(&block); block.call; end
                          3.349M (±15.1%) i/s -     32.290M

Comparison:
  def e1; yield; end:  4407918.3 i/s
def e2(&block); block.call; end:  3349074.7 i/s - 1.32x slower

Calculating -------------------------------------
def e3; e1 &Proc.new; end
                       100.244k i/100ms
def e4(&block); e1 &block; end
                       100.119k i/100ms
-------------------------------------------------
def e3; e1 &Proc.new; end
                          2.758M (±16.2%) i/s -     26.565M
def e4(&block); e1 &block; end
                          3.359M (±16.5%) i/s -     32.238M

Comparison:
def e4(&block); e1 &block; end:  3359376.3 i/s
def e3; e1 &Proc.new; end:  2757612.8 i/s - 1.22x slower
```
Results: There is overhead in MRI for creating procs (4.3x) as mentioned. In this particular case it creates the Proc anyway and `&Proc.new` is little slower in MRI and JRuby. JRuby is able to execute well both `yield` and `block.call` variants. I am inclining to following approach: not doing micro optimizations based on MRI. It would make the code more complicated and I am also assuming that users requiring performance will use JRuby anyway.